### PR TITLE
Add tag weights

### LINKED ISSUE #18: Build a tag cloud
Hello,
I wrote a plugin to create a tag cloud :
- [the plugin](https://framagit.org/lpaternault/ababsurdo/-/blob/9a8b6a5f37f3c13297bab375831c2c85b4582e98/packages/lektor-tag-cloud/lektor_tag_cloud.py)
- [the template producing the cloud](https://framagit.org/lpaternault/ababsurdo/-/blob/9a8b6a5f37f3c13297bab375831c2c85b4582e98/templates/blocks/tagcloud-block.html)
- [the resulting page](https://ababsurdo.fr/blog/tag/) (I suck at CSS, but you get the idea).

The plugin counts the number of articles for each tag, and attributes a "weight" to each tag, which can be a number (on [this page](https://ababsurdo.fr/blog/tag/), it is a number between 100 and 200 which is used as the text size), or a string (e.g. "small", "normal", "big", which would be used as classes), or whatever you want. The template itself is not included in the package.

Right now, it is a quick and dirty script in the git directory of the website using it. However, I would like to publish it properly.
- It is very simple and could be nicely integrated into `lektor-tags`. If you like this idea, I can write a pull request.
- If you prefer it to remain a separate plugin, I will clean it and publish it as a `lektor-tag-cloud` plugin in a few days or weeks.
Just tell me which one you want.

This could also be used to automatically generate a tags directory page (see #16).

Regards,
-- Louis

### PR DESCRIPTION
**Pull request by paternal**
Closes #18.

### CODE REVIEW

File: lektor_tags.py
Comment (paternal) [inline lektor_tags.py]: I'm a math teacher. Sorry, I couldn't help it… 😜 Feel free to remove it if not needed.

**Comment by paternal** [comment]
Almost done. The remaining features to add are (1) don't precalculate the tag cloud unless it's used somewhere and (2) if it's used, don't calculate it more than once per build.
I can do it by caching the tag count somewhere (where?). Then, when needed, if the cache exists, use it; if not calculate it. At the end (or beginning) of each build, remove cache. This would solve both points, but I do not know if it is the "lektor way" to do it? Is there already a mechanism implemented in lektor to cache plugin data?

@runfalk, @goanpeca ?

**Comment by paternal** [comment]
It works.

- `lektor build`: weights are not calculated, unless needed, and the results are cached, and they are calculated only once.
- `lektor serve`: same thing, but they are not calculated again if tags are changed. I see three solutions:
   - there exists an [event](https://www.getlektor.com/docs/api/plugins/events/) that is fired when lektor serve restarts a build process, and the cache is cleared there (but I could not see such an event);
   - we consider that `lektor serve` is only used for development purpose, and we explain this limitation in the documentation;
   - when `lektor serve` is used instead of `lektor build`, tags are never cached (and re-calculated for each new page where they are used).

What do you think?

File: lektor_tags.py
Comment (runfalk) [inline lektor_tags.py]: less -> least?

File: lektor_tags.py
Comment (runfalk) [reply lektor_tags.py]: You can probably also use the `most_common` method to replace min/max calculation as well.

File: lektor_tags.py
Comment (runfalk) [inline lektor_tags.py]: I know it's a bit more verbose (and less mathematical :sweat_smile:) but I think lower/upper_bound or just lower/upper are simpler to understand when looking at the signature.

If you want to you can still assign a = lower_bound when you use them.

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: > Couldn't this be renamed to `_weight_count`?

I have renamed it to `count()` (in the TagWeight class).

> I have not verified this, but will `self._tagcount` get cleared between rebuilds when Lektor runs in server mode (`lektor server`)?

No. See https://github.com/nixjdm/lektor-tags/pull/19#issuecomment-602286402. I will have another look at it, but later: my daughter is nagging at me right now…

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done as part of 1f3b141cd78935221395265978b0775944ee8aad

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done in 10b92138362cfac907c902943890919575550e1f

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done in 1e9d758e3284df9b3cc4504b21a7a0db8c4e7d08

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done in 53508fe751882a5efe4bfb06570be1e6133cb8b7

File: lektor_tags.py
Comment (runfalk) [reply lektor_tags.py]: I don't think it's that common in Python to such things. I never do such things unless I fork the project, and then I'll add a line to the LICENSE file. It seems like such a file is missing so @nixjdm probably has to add one. It seems the intended license is MIT (https://github.com/nixjdm/lektor-tags/blob/master/setup.py#L24)

Comment (paternal) [reply lektor_tags.py]: I removed my name here, and created files [AUTHORS](https://github.com/paternal/lektor-tags/blob/464f52cf06728c4b3e31f9d5a7b398b4d501b3b0/AUTHORS) and [LICENSE](https://github.com/paternal/lektor-tags/blob/464f52cf06728c4b3e31f9d5a7b398b4d501b3b0/LICENSE) files (based on the commit history.

File: lektor_tags.py
Comment (paternal) [inline lektor_tags.py:191]: This should be useless, since the very same line is present in the `on_before_build_all()` method, but `on_before_build_all()` is not always called before the very first build with `lektor server`. Should I report a bug on [lektor](http://github.com/lektor/lektor)?

**Comment by paternal** [comment]
Everything is OK for me.
@runfalk: Waiting for your review/merge. :smiley:

**Comment by runfalk** [comment]
@nixjdm would have to do that since this is his private project. I and other Lektor maintainers don't have access to this repo.

File: lektor_tags.py
Comment (runfalk) [reply lektor_tags.py:191]: Doesn't necessarily sound like a bug. Do you need the `on_before_build_all` hook?

Comment (paternal) [reply lektor_tags.py:191]: Yes. If it is not used, when using `lektor server`, TagWeights object is created in `on_setup_env()`, and is *never* updated (i.e. tag are counted once, even if tags are added/removed, they are not counted again). With `on_before_build_all()`, they are counted again at each build (but only once if they are needed several times in the same build).

Comment (runfalk) [reply lektor_tags.py:191]: I see, it's probably good to say that with a comment then. Like `# Needed to invalidate the tag cache before each build`

Comment (runfalk) [reply lektor_tags.py:191]: I assume it is not called for the first build with lektor server if there is nothing to be built? If there are other cases it does sound like a bug.

**Comment by paternal** [comment]
@nixjdm : Would you review this, or give @runfalk or @goanpeca (if they agree) the rights to accept this pull request and publish a new version on pypi?
Thanks

**Comment by paternal** [comment]
I am still interested in this pull request being merged. @yagebu, if I resolve conflicts, and ensure that it still works with latest versions of lektor and lektor-tags, would you review and maybe merge it?

This pull request counts tags (how many pages are tagged with "foo"?), and assign a weight to them. It can be used in templates to produce a list of tags (with number of occurrence), or produce a tag cloud ([this page](https://ababsurdo.fr/blog/tag/) is generated using the code of this pull request).

The changes in this pull request [are documented](https://github.com/lektor/lektor-tags/pull/19/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5).

Regards,
paternal

**Comment by dairiki** [comment]
My 2¢ on API: Have you considered, instead of `TagWeights` providing a bunch of methods that produce mappings, making the `TagWeights` instance be a mapping that maps tag names to `TagWeight` objects, which, in turn, provide methods or attributes that provide the various weights.

This seems cleaner to me, and would make it easier to do things that are interested in multiple weight-types at the same time, like:

```jinja
<ul>
{% for tag, weight in tagweights %}
  <li style="{{ weight.log(100, 200)|round|int }}%;">
    {{ tag }} ({{ weight.count }} articles).
  </li>
{% endfor %}
</ul>
```

The `TagWeight` values would be orderable (by count), so ordering tags via `dictsort` would still work.

**Comment by paternal** [comment]
> My 2¢ on API: Have you considered, instead of `TagWeights` providing a bunch of methods that produce mappings, making the `TagWeights` instance be a mapping that maps tag names to `TagWeight` objects, which, in turn, provide methods or attributes that provide the various weights.
> 
> […]

I agree 95e0f25

**Review by runfalk — APPROVED** [review]

**Comment by runfalk** [comment]
The changes look good to me but I haven't tested anything. If @dairiki also think it looks OK I think we can merge

**Review by dairiki — COMMENTED** [review]
It would also be good to have some unit tests for the new code.  (Though, I realize the state of the existing tests is not great.)

I've pushed a proposed `tests/test_tagweights.py` to your branch.  (I hope you don't mind.)
(One test currently fails due to `KeyError` in `TagsPlugin.tagcount` as described in another comment.)

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py:74]: Done 8d010e9c2dcfae7699eb9302df134ce4b54b95cc

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done d2b68999648ad973614e8ff6c266ed85ed0c67d7

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done 6be1c9d62854cbf2083a6cf72a1b3f9cc04213de

Note: ``counter.total()`` appeared in python3.10: you are ahead of your time!

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done 91f5da166eb8dc29efd49dc7e174d1cbdee9de90

I rewrote `TagWeight.linear()` the same way.

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done e1cfa919138deb7840efa9cb3a58a135ac6cbe38

Well spotted! In early versions of this pull request, the weights were *computed* here, and needed to be computed again at each build. This is indeed no longer necessary.

File: lektor_tags.py
Comment (paternal) [reply lektor_tags.py]: Done cf9753a7f67c34e66db01625193b25ace48b93e9

**Comment by paternal** [comment]
> I've pushed a proposed `tests/test_tagweights.py` to your branch. (I hope you don't mind.)

No problem. Thanks! :)

File: lektor_tags.py
Comment (dairiki) [reply lektor_tags.py]: Good catch.

**Comment by dairiki** [comment]
Merged and released as lektor-tags 0.5.0.

Thank you @paternal!

**Comment by paternal** [comment]
Thank you for your careful reviews!

**Pull request merged**
